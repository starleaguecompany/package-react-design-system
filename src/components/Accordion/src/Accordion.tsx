import * as React from 'react'
import { useStyles, useMultiValue, safeInvoke, isNull } from '@starleaguecompany/package-react-utils'

import { Space } from '../../Space'

import { AccordionProps } from '../types/Accordion.types'
import styles from '../styles/Accordion.module.less'

export * from '../types/Accordion.types'

/**
 * @description A content area which can be collapsed and expanded
 *
 * @component
 * @example
 * ```jsx
 * <Accordion activeKeys={['credit']}>
 *   <Accordion.Item key="credit" title="О кредите">content</Accordion.Item>
 * </Accordion>
 * ```
 */
const Accordion = React.forwardRef<HTMLDivElement, AccordionProps>((props, ref) => {
  const { defaultIndex, separator, background, scrollIntoView, children, className, onChange, ...restProps } = props
  const [innerValue, setInnerValue] = useMultiValue<number>(defaultIndex || [])

  const cx = useStyles(styles)

  const classNames = cx(className, {
    [`separator-${separator}`]: separator,
    [`background-${background}`]: background,
  })

  const handleCollapse = React.useCallback(
    event => {
      const value = Number(event.currentTarget.dataset.index)

      setInnerValue(value, val => {
        safeInvoke(onChange, val)
      })
    },
    [innerValue]
  )

  const separatorSpace = separator === 'space' ? 16 : 0

  const panels = React.useMemo(() => {
    return React.Children.map(children as React.ReactElement, (child, index) => {
      if (isNull(child)) return

      const childProps = {
        ...child.props,
        id: index,
        active: innerValue.indexOf(index) >= 0,
        scrollIntoView,
        onCollapse: handleCollapse,
      }

      return React.cloneElement(child, childProps)
    })
  }, [innerValue, children])

  return (
    <Space
      ref={ref}
      data-qa="Accordion"
      size={separatorSpace}
      direction="vertical"
      className={classNames}
      {...restProps}
    >
      {panels}
    </Space>
  )
})

Accordion.defaultProps = {
  defaultIndex: [],
  separator: 'space',
  scrollIntoView: true,
}

export default Accordion
