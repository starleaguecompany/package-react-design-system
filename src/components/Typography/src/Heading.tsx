import * as React from 'react'
import { useStyles } from '@starleaguecompany/package-react-utils'

import { HeadingProps } from '../types/Heading.types'
import styles from '../styles/Heading.module.less'

/**
 * @description The Heading component is used for all headings.
 *
 * @component
 * @example
 * ```jsx
 * <Heading level={2}>
 *   Heading
 * </Heading>
 * ```
 */
const Heading = React.forwardRef<HTMLHeadingElement, HeadingProps>((props, ref) => {
  const { as, level, className, children, ...restProps } = props
  const cx = useStyles(styles)

  const classNames = cx(className, 'container', {
    [`size-${level}`]: level,
  })

  return React.createElement(
    as || (`h${level}` as React.ElementType),
    {
      ref: ref,
      className: classNames,
      ...restProps,
    },
    children
  )
})

Heading.defaultProps = {
  level: 1,
}

export { Heading }
