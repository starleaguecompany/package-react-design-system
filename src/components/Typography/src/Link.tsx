import * as React from 'react';
import { useStyles } from '@starleaguecompany/package-react-utils';

import { LinkProps } from '../types/Link.types';
import styles from '../styles/Link.module.less';

/**
 * @description The Link component is used for anchor links. This component renders a a element by default
 *
 * @component
 * @example
 * ```jsx
 * <Link href="#">Link</Link>
 * ```
 */
const Link = React.forwardRef<HTMLAnchorElement, LinkProps>((props, ref) => {
  const { color, strong, className, children, ...restProps } = props;
  const cx = useStyles(styles);

  const classNames = cx(className, 'container', {
    [`color-${color}`]: color,
    strong,
  });

  return (
    <a ref={ref} className={classNames} {...restProps}>
      {children}
    </a>
  );
});

Link.defaultProps = {
  color: 'blue',
};

export { Link };
