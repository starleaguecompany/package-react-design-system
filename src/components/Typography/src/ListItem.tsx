import * as React from 'react'
import { useStyles } from '@starleaguecompany/package-react-utils'

import { ListItemProps } from '../types/ListItem.types'
import styles from '../styles/List.module.less'

/**
 * @description List item element.
 *
 * @component
 * @example
 * ```jsx
 * <ListItem>...</ListItem>
 * ```
 */
const ListItem = React.forwardRef<HTMLLIElement, ListItemProps>((props, ref) => {
  const { icon, className, children, ...restProps } = props
  const cx = useStyles(styles)

  const classNames = cx(className, 'item', {
    'with-icon': icon,
  })

  return (
    <li ref={ref} className={classNames} {...restProps}>
      {icon ? <span className={cx('icon')}>{icon}</span> : null}
      {children}
    </li>
  )
})

export { ListItem }
