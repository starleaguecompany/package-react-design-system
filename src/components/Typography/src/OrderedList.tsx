import * as React from 'react';
import { useStyles } from '@starleaguecompany/package-react-utils';

import { OrderedListProps } from '../types/OrderedList.types';
import styles from '../styles/List.module.less';

/**
 * @description Ordered list element.
 *
 * @component
 * @example
 * ```jsx
 * <OrderedList>...</OrderedList>
 * ```
 */
const OrderedList = React.forwardRef<HTMLOListElement, OrderedListProps>((props, ref) => {
  const { className, children, ...restProps } = props;
  const cx = useStyles(styles);

  const classNames = cx(className, 'ordered');

  return (
    <ol ref={ref} className={classNames} {...restProps}>
      {children}
    </ol>
  );
});

export { OrderedList };
