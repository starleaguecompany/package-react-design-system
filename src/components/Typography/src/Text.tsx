import * as React from 'react';
import { useStyles } from '@starleaguecompany/package-react-utils';

import { FONT_SIZES } from '../../../constants/sizes';

import { TextProps } from '../types/Text.types';
import styles from '../styles/Text.module.less';

/**
 * @description The Text component is used for single line or multiline text. The component renders a div element by default
 *
 * @component
 * @example
 * ```jsx
 * <Text as="div" size="xl" strong>Text</Text>
 * ```
 */
const Text = React.forwardRef<HTMLDivElement, TextProps>((props, ref) => {
  const {
    as,
    size,
    strong,
    uppercase,
    capitalize,
    lowercase,
    truncate,
    nowrap,
    lineThrough,
    className,
    children,
    ...restProps
  } = props;
  const cx = useStyles(styles);

  const classNames = cx(className, {
    [`size-${size}`]: true,
    strong,
    uppercase,
    capitalize,
    lowercase,
    truncate,
    nowrap,
    lineThrough,
  });

  return React.createElement(
    as as React.ElementType,
    {
      ref: ref,
      className: classNames,
      ...restProps,
    },
    children
  );
});

Text.defaultProps = {
  size: FONT_SIZES.S14,
  as: 'div',
};

export { Text };
