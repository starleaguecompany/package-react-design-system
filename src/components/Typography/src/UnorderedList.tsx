import * as React from 'react';
import { useStyles } from '@starleaguecompany/package-react-utils';

import { UnorderedListProps } from '../types/UnorderedList.types';
import styles from '../styles/List.module.less';

/**
 * @description Unordered list element.
 *
 * @component
 * @example
 * ```jsx
 * <UnorderedList>...</UnorderedList>
 * ```
 */
const UnorderedList = React.forwardRef<HTMLUListElement, UnorderedListProps>((props, ref) => {
  const { className, children, ...restProps } = props;
  const cx = useStyles(styles);

  const classNames = cx(className, 'unordered');

  return (
    <ul ref={ref} className={classNames} {...restProps}>
      {children}
    </ul>
  );
});

export { UnorderedList };
