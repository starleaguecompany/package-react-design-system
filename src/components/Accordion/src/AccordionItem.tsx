import * as React from 'react'
import { noop, useStyles } from '@starleaguecompany/package-react-utils'
import { ArrowDown } from '@starleaguecompany/react-icons'

import { Icon } from '../../Icon'
import { Collapse } from '../../Transition'
import { Space } from '../../Space'

import { AccordionItemProps } from '../types/Accordion.types'
import styles from '../styles/Accordion.module.less'

/**
 * @description Accordion Item component.
 *
 * @component
 * @example
 * ```jsx
 * <Accordion.Item key="credit" title="О кредите">content</Accordion.Item>
 * ```
 */
const Item = React.forwardRef<HTMLDivElement, AccordionItemProps>((props, ref) => {
  const { id, title, subtitle, icon, active, scrollIntoView, tabIndex, className, children, onCollapse, ...restProps } =
    props
  const activeItemRef = React.useRef<HTMLDivElement | null>(null)

  const cx = useStyles(styles)

  const collapse = React.useCallback(
    (event: React.MouseEvent<HTMLDivElement> | React.KeyboardEvent<HTMLDivElement>) => {
      if (scrollIntoView) {
        activeItemRef.current = event.currentTarget.closest('[data-qa=CollapseItem]')

        setTimeout(() => {
          activeItemRef.current && activeItemRef.current.scrollIntoView({ behavior: 'smooth', block: 'nearest' })
        }, 300)
      }

      onCollapse && onCollapse(event)
    },
    [onCollapse, scrollIntoView]
  )

  const handleCollapse = (event: React.MouseEvent<HTMLDivElement>) => {
    collapse(event)
  }

  const handleKeypress = (event: React.KeyboardEvent<HTMLDivElement>) => {
    if (event.key === 'Enter') {
      event.preventDefault()
      collapse(event)
    }
  }

  return (
    <div ref={ref} data-qa="CollapseItem" className={cx(className, 'panel', { active })} {...restProps}>
      <div
        className={cx('header', { active })}
        data-index={id}
        tabIndex={tabIndex}
        onClick={children ? handleCollapse : noop}
        onKeyPress={handleKeypress}
      >
        <Space size={12} align="center">
          {icon && <div className={cx('icon')}>{icon}</div>}
          {title || subtitle ? (
            <Space direction="vertical" size={2} className={cx('titleWrapper')}>
              {title && <div className={cx('title')}>{title}</div>}
              {subtitle && <div className={cx('subtitle')}>{subtitle}</div>}
            </Space>
          ) : null}
        </Space>
        {children && <Icon className={cx('collapseIcon')} icon={<ArrowDown />} />}
      </div>
      <Collapse visible={active} className={cx('content')} unmountOnExit>
        <div className={cx('content-wrapper')}>{children}</div>
      </Collapse>
    </div>
  )
})

Item.defaultProps = {
  tabIndex: 0,
}

export default Item
