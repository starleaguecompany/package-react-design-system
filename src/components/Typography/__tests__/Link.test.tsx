import * as React from 'react'
import { render, screen } from '@testing-library/react'

import { Link } from '../src/Link'

describe('Link', () => {
  test('should render correctly with no props', () => {
    const { container } = render(<Link />)

    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toHaveClass('container color-blue', { exact: true })
  })

  test('should render correctly with text', () => {
    render(<Link href="#">Testing link</Link>)

    expect(screen.getByRole('link')).toMatchSnapshot()
    expect(screen.getByRole('link', { name: 'Testing link' })).toBeVisible()
  })

  test('should render correctly with color prop', () => {
    render(
      <Link href="#" color="black">
        Testing link
      </Link>
    )

    expect(screen.getByRole('link')).toMatchSnapshot()
    expect(screen.getByRole('link', { name: 'Testing link' })).toHaveClass('color-black')
  })

  test('should render correctly with strong prop', () => {
    render(
      <Link href="#" strong>
        Testing link
      </Link>
    )

    expect(screen.getByRole('link')).toMatchSnapshot()
    expect(screen.getByRole('link', { name: 'Testing link' })).toHaveClass('strong')
  })

  test('should render correctly with attributes', () => {
    render(
      <Link id="test-id" className="test-class" href="https://test.ru">
        Testing link
      </Link>
    )

    expect(screen.getByRole('link', { name: 'Testing link' })).toMatchSnapshot()
    expect(screen.getByRole('link', { name: 'Testing link' })).toHaveClass('test-class')
    expect(screen.getByRole('link', { name: 'Testing link' })).toHaveAttribute('id', 'test-id')
    expect(screen.getByRole('link', { name: 'Testing link' })).toHaveAttribute('href', 'https://test.ru')
  })
})
