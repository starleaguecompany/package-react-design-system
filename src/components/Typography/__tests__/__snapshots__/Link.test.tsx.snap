// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render correctly with attributes 1`] = `
<a
  class="test-class container color-blue"
  href="https://www.sravni.ru"
  id="test-id"
>
  Testing link
</a>
`;

exports[`Link should render correctly with color prop 1`] = `
<a
  class="container color-black"
  href="#"
>
  Testing link
</a>
`;

exports[`Link should render correctly with no props 1`] = `
<a
  class="container color-blue"
/>
`;

exports[`Link should render correctly with strong prop 1`] = `
<a
  class="container color-blue strong"
  href="#"
>
  Testing link
</a>
`;

exports[`Link should render correctly with text 1`] = `
<a
  class="container color-blue"
  href="#"
>
  Testing link
</a>
`;
